lib/pos_transaction_register.go

```diff
@@ -274,12 +274,10 @@ func (tr *TransactionRegister) Reset() {
 	tr.totalTxnsSizeBytes = 0
 }
 
-// GetFeeTimeIterator returns an iterator over the transactions in the register. The iterator goes through all transactions
-// as ordered by Fee-Time.
+// GetFeeTimeIterator returns an iterator over the transactions in the register. The iterator goes through all
+// transactions as ordered by Fee-Time. The caller must hold a read lock on the transaction register before
+// calling this function and while iterating over the transactions.
 func (tr *TransactionRegister) GetFeeTimeIterator() *FeeTimeIterator {
-	tr.RLock()
```

Comment:
Calling `tr.RLock()` internally here while holding `tr.RLock()` inside `PosFeeEstimator.estimateFeeRateNanosPerKBGivenTransactionRegister` is one of the things that causes the deadlock.

In general, it doesn't make sense to grab and release the read lock here cause the caller needs to hold the lock the whole time that the iterator is active and can be used. That means the caller needs to hold the lock after this function has returned. The lock hold here is redundant.